File: versa_system/public/js/work_order.js
frappe.ui.form.on("Work Order", {
  onload: function (frm) {
    // Add the custom button only if the Sales Order is linked
    if (frm.doc.sales_order && frm.doc.status === "Completed") {
      updateAndNavigateToSalesOrder(frm);
    }
  },

  // Triggered when the status is updated or any changes occur
  status: function (frm) {
    // Add the button if the status is 'Completed'
    if (frm.doc.status === "Completed") {
      updateAndNavigateToSalesOrder(frm);
    } else {
      frm.remove_custom_button(__("Go to Sales Order"));
    }
  },
});

// Function to add button and navigate to Sales Order form
function updateAndNavigateToSalesOrder(frm) {
  // Add a custom button to navigate to the Sales Order form
  frm.add_custom_button(
    __("Go to Sales Order"), // Button label
    function () {
      if (!frm.doc.sales_order) {
        frappe.msgprint(__("Please select a Sales Order before proceeding."));
        return;
      }

      // Update the Sales Order status to "Proforma Invoice"
      frappe.call({
        method:
          "versa_system.versa_system.custom_scripts.work_order.update_sales_order_status_on_work_order_completion",
        args: {
          work_order_name: frm.doc.name,
        },
        callback: function (response) {
          if (response.message === "Success") {
            frappe.msgprint(
              __("Sales Order status updated to 'Proforma Invoice'.")
            );
            // Navigate to the Sales Order form using the linked Sales Order
            frappe.set_route("Form", "Sales Order", frm.doc.sales_order);
          } else {
            frappe.msgprint(__("Error: ") + response.message);
          }
        },
        error: function (error) {
          frappe.msgprint(
            __("Error occurred while updating the Sales Order: ") +
              error.message
          );
        },
      });
    },
    __("Create") // Button group (default is "Create" group)
  );
}

File: versa_system/versa_system/custom_scripts/work_order.py
import frappe

@frappe.whitelist()
def update_sales_order_status_on_work_order_completion(work_order_name):
    """Update the status of the linked Sales Order when Work Order is completed."""
    try:
        frappe.log_error(f"update_sales_order_status_on_work_order_completion called for Work Order: {work_order_name}")
        
        # Fetch the Work Order document
        work_order = frappe.get_doc("Work Order", work_order_name)

        # Ensure Work Order is linked to a Sales Order
        if work_order.sales_order:
            # Fetch the Sales Order document
            sales_order = frappe.get_doc("Sales Order", work_order.sales_order)
            frappe.log_error(f"Work Order Status: {work_order.status}, Sales Order Status: {sales_order.status}")
            
            # Update Sales Order status if Work Order is completed
            if work_order.status == "Completed":
                frappe.db.set_value("Sales Order", sales_order.name, "status", "Proforma Invoice")
                frappe.db.commit()
                frappe.log_error(f"Sales Order status updated to 'Proforma Invoice'.")
                return "Success"
            else:
                frappe.log_error(f"Work Order status is not 'Completed'.")
                return "Work Order is not 'Completed'"
        else:
            frappe.log_error(f"No Sales Order linked to Work Order {work_order_name}.")
            return "No Sales Order linked to Work Order"

    except frappe.DoesNotExistError:
        frappe.log_error(f"Work Order {work_order_name} does not exist.")
        return "Work Order does not exist"
    except Exception as e:
        frappe.log_error(f"Failed to update Sales Order status: {str(e)}")
        return f"Error: {str(e)}"
